Use classes for plugins

lafrech opened the pull request:
Closes #207.

I think I'm almost done with this.

There is still a few things to do:

- [x] Decide whether or not to provide a base class for plugin classes. @sloria, you seemed to prefer not to, but I don't see why. I think it would simplify plugin creation.
- [x] Be sure to complete tests to test Plugin interface
- [x] Document changes

lafrech committed:
Use classes for plugins

lafrech committed:
Plugin: make file level functions plugin object methods

lafrech committed:
Remove useless import in test_swagger.py

lafrech committed:
Move schema_name_resolver to MarshmallowPlugin

lafrech committed:
Store schema refs in Swagger object

lafrech committed:
Add ext.mashmallow.swagger.Swagger

lafrech committed:
Re-introduce register_*_helper methods

lafrech committed:
Rework APISpec plugins management

lafrech committed:
Backward compatibilty with old plugin interface for bottle/flask/tornado

lafrech committed:
Backward compatibilty with old plugin interface for marshmallow ext

lafrech committed:
Marshmallow swagger ext: don't pass spec object through all functions

lafrech committed:
Reintroduce plugin setup tests

lafrech committed:
Fix marshmallow/swagger tests

lafrech committed:
Remove unused import

lafrech committed:
MarshmallowPlugin: behave according to openapi_version

lafrech committed:
Bugfix: OpenAPI 2.0.0 does not exist. Use 2.0

lafrech committed:
Use warnings.catch_warning context manager to test warnings

lafrech committed:
Add ext/marshmallow/common.py

common.py contains resolve_schema_instance and resolve_schema_cls.

This commit removes a duplication (get_schema_class / resolve_schema_cls)
and solves a circular import issue.

lafrech committed:
marshmallow.swagger: Make FIELD_MAPPING an attribute of Swagger

lafrech committed:
Centralize spec fixtures in conftest

lafrech committed:
Rework openapi version management

Put validate_openapi_version in utils.py
Let it accept either string or LooseVersion representation
Call it from classes taking OpenAPI version as parameter

lafrech committed:
Introduce utils.OpenAPIVersion

lafrech committed:
Add deprecation warnings for old plugin/helpers interface

lafrech committed:
Rework MarshmallowPlugin init

Remove openapi_version parameter
Don't instantiate Swagger until init_spec

lafrech committed:
Introduce BasePlugin: base class for plugins

sloria commented on apispec/core.py:
```diff
@@ -125,24 +124,38 @@ def __init__(self, title, version, plugins=(), info=None,
+        old_plugins = list(p for p in plugins if isinstance(p, str))
+        if old_plugins:
+            warnings.warn(
+                'Old style plugins are deprecated. Use classes instead.', DeprecationWarning)
```
Note for the future: it'd be nice if this linked to the relevant docs.

```
Old-style plugins are deprecated. Use plugin classes instead. See https://apispec.readthedocs.io/en/latest/using_plugins.html.
```

sloria commented on apispec/core.py:
```diff
@@ -319,6 +366,8 @@ def register_definition_helper(self, func):
         :param callable func: The definition helper function.
         """
+        warnings.warn(
+            'Helper functions are deprecated. Use plugin classes.', DeprecationWarning)
```
Same here: once docs are written, you can link to them:

```
Helper functions are deprecated. Use plugin classes instead. See https://apispec.readthedocs.io/en/latest/writing_plugins.html
```

sloria reviewed (commented):
I like the direction this is going. I don't see any major design issues.

I think you're right about the base plugin class simplifying plugin creation, since there's shared initialization logic across plugins. It would also make the plugin interface more "self-documenting". IDEs would be able to do static analysis, etc.

lafrech committed:
Merge branch 'dev' into dev_207_plugin_classes

lafrech commented:
I moved `BasePlugin` out of `ext` and made it importable from `apispec`.

I also rebased. I hope to get this done and merged soon because the change is so important every rebase is a nightmare. In fact, I had to merge rather than rebase as the rebase was too complicated.

lafrech committed:
Rework test_core.py for better v2/v3 parametrization

lafrech committed:
Remove BasePlugin.__init__(). Plugins must be passed to APISpec at init.

lafrech committed:
BasePlugin: make path and operations keyword args in path/operation helpers

lafrech committed:
Tornado path helper: treat {} operations as None

lafrech committed:
core.add_path plugin interface: rework, test, add docstrings

- Deal with operations consistently whether or not path parameter is a
string, a Path without operations or a Path with operations.
- Fix case where operation helpers mutate operations local variable but
not path.operations.
- Allow plugin response helpers to return None.

lafrech committed:
core.add_path: check path.path before calling operation helpers

lafrech commented:
I added tests for the plugin interface, and docstrings in the `BasePlugin` class to explain what the helper methods should do.

I also reworked `add_path` a bit to fix some inconsistencies. Since this is a huge rework, it is tempting to change everything. I'm still not perfectly happy about the whole thing but this is out of scope so I'd rather postpone to later (potentially breaking) changes. For instance the `path` and `operations` parameters passed from `add_path` to the plugin helpers are redundant as operations is an attribute of path. Also, I don't find the API really consistent. Some helpers return values while some mutate their inputs, some may return `None`, some must return a `dict`... I tried to stay close to current interface.

The last missing part is the documentation.

I'm happy if I get some feedback before writing the docs so that I don't have to rewrite if some changes are needed in the code.

There is no "upgrading" section in the docs. May I stick to a few lines in the changelog explaining what to do when upgrading (along with a rewrite of the "writing plugins" section)?

sloria commented:
Re: API inconsistencies: I don't doubt there's some things to be cleaned up. But those don't necessarily need to block getting this merged, especially if they're risky to change. As for low-risk changes: feel free to clean things up as you see fit.

I'll try to review this more over the weekend if I have time, though I can't make any promises (I'm currently in the process of moving between houses).

> There is no "upgrading" section in the docs. May I stick to a few lines in the changelog explaining what to do when upgrading (along with a rewrite of the "writing plugins" section)?

Yep, that's perfect.

lafrech committed:
Merge commit '27a853b' (dev branch) into dev_207_plugin_classes

lafrech committed:
Merge commit 'f0f7fbc' (dev branch) into dev_207_plugin_classes

lafrech committed:
Merge branch 'dev' into dev_207_plugin_classes

lafrech commented:
Rebased again.

I really hope to merge this soon because each merge is a PITA. Not only because there are so many changes, but also because this PR improves the test coverage quite a lot: all tests (or almost) are run against both OpenAPI v2 and v3 (while in dev, only a few specific tests are run against v3).

The difficulty here was not so much the merge itself, but fixing two tricky bugs that had just been added to dev that the tests in dev wouldn't show. Those are fixed in the merge commits, which is not so nice in terms of traceability.

I'm not so happy with `fields2parameters`. It has `default_in` defaulting to `body` which is invalid in OpenAPIv3, and its internal logic is a bit twisted. It might be worth splitting in two methods for v2 and v3. Anyway, this can be reworked later on.

sloria commented on apispec/ext/marshmallow/swagger.py:
```diff
@@ -48,211 +50,15 @@
 CUSTOM_FIELD_MAPPING_ATTR = '__swagger_field_mapping'
 
 
-def map_to_swagger_type(*args):
```
This was part of the public API. Perhaps this should be exposed `MarshmallowPlugin` as an instance method?

lafrech replied:
It is already exposed. Sort of. It is in the `Swagger` object.

```python
ma_plugin  = MarshmallowPlugin()

APISpec(
        title='My super API',
        version='0.1',
        plugins=(ma_plugin, ),
        openapi_version='3.0.1')

@ma_plugin.swagger.map_to_swagger_type('string', 'uuid')
class UuidField(StrField):
    [...]
```

I could add a `map_to_swagger_type` method to `MarshmallowPlugin` to proxy this one, but unless I add complicated or awkward logic, it would still have to be called after registration in `APISpec` instance, as this is when the `Swagger` instance is created.

I can't think of any simpler way to do this. We could create the `Swagger` object at init of the MA plugin, but it still wouldn't be totally usable until a spec object is set (just like the MA plugin) anyway.

Or maybe we can merge the MA plugin and the swagger object. Using composition, we'd keep them in separate files.

```python
class MarshmallowPlugin(BasePlugin, Swagger):
    """APISpec plugin handling marshmallow schemas
    [...]
    """
```

This way, all methods are accessible at MA plugin level.

Looks weird to me, but perhaps because I too rarely use composition.

sloria commented on apispec/core.py:
```diff
@@ -213,35 +227,61 @@ def normalize_path(path):
+        for plugin in self.plugins:
+            try:
+                plugin.operation_helper(path=path, operations=path.operations, **kwargs)
+            except NotImplementedError:
```
It's possible for this to swallow `NotImplementedError` errors that aren't raised directly by plugin methods. I suppose `BasePlugin` could raise a plugin-specific exception, e.g. `PluginMethodNotImplementedError` to avoid this. Might not be absolutely necessary, though.

lafrech replied:
Yes, I'm not sure sure it is that bad. I'm more uncomfortable with the `TypeError` caught above. I already experienced issues because it was hiding real problems in my code. `NotImplementedError` is more specific. But I don't mind changing this if you wish.

sloria replied:
Yeah, I'm also concerned about TypeErrors being swallowed. I don't recall why that exception-handling was added. Would have to dig more into it.

lafrech replied:
Perhaps was it made to avoid breaking on helper functions that would not have **kwargs in their signature?

I could remove it while I'm at it (no test broken on this branch if I do).

sloria replied:
Ah, that seems likely. Yeah, go ahead and remove it.

lafrech replied:
OK, I removed `TypeError` catching for the new plugin interface and left it on the legacy helpers.

I also introduced `PluginMethodNotImplementedError`.

sloria reviewed (commented):
This is looking great. Thanks for all your work keeping this up to date.

Before we release this change, I think `swagger` and `Swagger` should be renamed `open_api` and `OpenAPIConverter`. But that doesn't need to happen before this PR is merged.

lafrech committed:
Swagger: remove unused CUSTOM_FIELD_MAPPING_ATTR

lafrech committed:
Rename Swagger stuff as OpenAPI

lafrech commented:
> Before we release this change, I think `swagger` and `Swagger` should be renamed `open_api` and `OpenAPIConverter`.

Done. Except I chose `openapi` over `open_api`. This is the name of the version field in the spec (https://swagger.io/specification/#oasVersion).

lafrech committed:
Deprecate APISpec.setup_plugin()

lafrech commented:
Only `MarshmallowPlugin` uses the spec object.

I'm thinking of emptying `BasePlugin.init_spec` and store spec object as `spec` attribute only in `MarshmallowPlugin.init_spec`.

lafrech committed:
BasePlugin: raise PluginMethodNotImplementedError rather than NotImplementedError

Also, don't catch TypeError from plugin methods in APISpec.

lafrech committed:
Document plugin class interface

lafrech commented:
Documentation updated.

TODO:

- [ ] Decide if `spec` should be stored as attribute in `BasePlugin.init_spec`. We can keep it as it is. No big deal.
- [x] Changelog. Should I wait until it is merged into `dev`?
- [ ] Perhaps add some `.. deprecated:: 0.38.0` directive in every plugin module's `setup` function. Version number unkown until merged into `dev`.

lafrech committed:
Update CHANGELOG

lafrech commented:
CHANGELOG updated.

Also, I'm hesitant to make `[bottle|flask]_path_to_swagger/openapi` Plugin class static method or leave them as global module functions. That's a minor detail.

@sloria, I think I'm done with this for now. Thanks (and good luck !) for the review.

sloria commented:
>  Decide if spec should be stored as attribute in BasePlugin.init_spec. We can keep it as it is. No big deal.

I suppose making `init_spec` no-op makes plugin classes more general-purpose. I don't have strong feelings about it either way, though. 

>  Perhaps add some .. deprecated:: 0.38.0 directive in every plugin module's setup function. Version number unkown until merged into dev.

Good idea. Feel free to assume 0.39.0 (i.e. `.. deprecated:: 0.39.0`) for now. I can always change those before I release.

> Also, I'm hesitant to make [bottle|flask]_path_to_swagger/openapi Plugin class static method or leave them as global module functions. That's a minor detail.

Yeah, it's personal preference. I tend to eschew static methods in favor of functions.

> I think I'm done with this for now. Thanks (and good luck !) for the review.

Great! I'll try to give this a final pass some time within the next week.

lafrech committed:
Don't store spec object in BasePlugin.init_spec

lafrech committed:
Update documentation

lafrech commented:
> I suppose making init_spec no-op makes plugin classes more general-purpose.

Done.

> add some .. deprecated:: 0.39.0 directive in every plugin module's setup function.

Done. 

> make [bottle|flask]_path_to_swagger/openapi Plugin class static method or global module functions

Left as static methods. I like to have all methods accessible from the class or object, rather than "scattered" in the module. You're right, that's a matter of personal preference.

> I'll try to give this a final pass some time within the next week.

Great, thanks.

sloria commented on apispec/exceptions.py:
```diff
@@ -4,16 +4,15 @@
     """Raised when a plugin cannot be found or is invalid."""
-    pass
+
+class PluginMethodNotImplementedError(APISpecError):
```
Minor: This should probably inherit from `NotImplementedError`.

lafrech replied:
Done.

```python
class PluginMethodNotImplementedError(APISpecError, NotImplementedError):
    """Raised when calling an unimplemented helper method in a plugin"""
```

sloria commented on docs/using_plugins.rst:
```diff
@@ -195,18 +193,33 @@ If your API uses `method-based dispatching <http://flask.pocoo.org/docs/0.12/vie
+    )
+
+
+    @ma_plugin.openapi.map_to_swagger_type('string', 'uuid')
```
This breaks the [Law of Demeter](https://www.wikiwand.com/en/Law_of_Demeter) by accessing a method on an internal attribute. Also, it's a lot of typing. Perhaps we should expose `map_to_swagger_type` on `MarshmallowPlugin`?

lafrech replied:
Done.

sloria commented on docs/writing_plugins.rst:
```diff
@@ -3,12 +3,15 @@
 
-Helper Functions
-----------------
+A plugins is a child class of `apispec.plugin.BasePlugin`.
```
Revision: "A plugin is a subclass of `apispec.plugin.BasePlugin`."

sloria commented on docs/writing_plugins.rst:
```diff
@@ -17,67 +20,73 @@ There are four types of helper functions:
+The ``init_spec`` Method
+------------------------
 
+`BasePlugin` has an `init_spec` method that `APISpec` calls on each plugin at initialization with the spec object itself as parameter. It is a no-op, but it can be overriden by a plugin to get useful information from the spec object.
```
Revision suggestion: "It is no-op by default, but a plugin may override it to access and store useful information from the spec object."

sloria commented on docs/writing_plugins.rst:
```diff
@@ -17,67 +20,73 @@ There are four types of helper functions:
+            """Path helper that parses docstrings for operations. Adds a
+            ``func`` parameter to `apispec.APISpec.add_path`.
+            """
+            openapi_major_version = self.spec.openapi_version.major
```
This would require `spec` to be stored in `init_spec`.

sloria commented on tests/conftest.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@pytest.fixture(params=(
+    ('2.0', True),
```
Nice! 👍

sloria reviewed (changes requested):
This looks excellent. Just a few minor changes suggested, but otherwise this looks good to go.

lafrech committed:
Expose OpenAPIConverter.map_to_openapi_type from MarshmallowPlugin

lafrech commented:
Thanks for the review. That's a huge change. We may have left a few glitches. Nothing big, hopefully.

I addressed all your comments.

sloria commented:
I'll cut a release after work today.

lafrech commented:
No rush. Thanks.
